### PR TITLE
Explicit binders in cases

### PR DESCRIPTION
**Pull request by munificent**
OK, I've revamped the patterns proposal to resolve (hopefully once and for all!) how we want to handle variable binders.

I think this is overall a significant improvement over the current proposal. It's roughly similar to the original proposal, but still retains the unified grammar in terms of how its specified and is, I think, easier to understand.

cc @stereotype441 @natebosch @jakemac53 @kallentu 

Note that there is a "CHANGELOG" section at the end of the proposal that goes into more detail about what actually changed.

### CODE REVIEW

File: working/0546-patterns/patterns-feature-specification.md [hunk 1]
```diff
@@ -1885,14 +1900,27 @@ Where `<keyword>` is `var` or `final` is treated like so:
...
 
-#### Pattern-if statement
+#### If-case statement
 
 1.  Evaluate the `expression` producing `v`.
 
-2.  Match the `pattern` against `v`.
+2.  Match the `pattern` in the `caseHead` against `v`.
+
+3.  If the match succeeds:
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1909]: Worth mentioning that new variables are available here?

File: working/0546-patterns/patterns-feature-specification.md [hunk 2]
```diff
@@ -1885,14 +1900,27 @@ Where `<keyword>` is `var` or `final` is treated like so:
...
+
+    1.  If there is a guard clause:
+
+        1.  Evaluate it. If it does not evaluate to a Boolean, throw a runtime
+            exception. *This can happen if the guard expression's type is
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1914]: exception -> error.

In practice, this should be a completely normal failing implicit downcast from `dynamic` to `bool`.

File: working/0546-patterns/patterns-feature-specification.md [hunk 3]
```diff
@@ -1885,14 +1900,27 @@ Where `<keyword>` is `var` or `final` is treated like so:
...
+            `dynamic`.*
 
-3.  If the match succeeds, evaluate the then `statement`. Otherwise, if there
-    is an `else` clause, evaluate the else `statement`.
+        1.  If the guard evaluates to `true`, evaluate the then `statement`.
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1917]: Nit: evaluate -> execute (which is what we say we do for statements, evaluation is for expressions).

File: working/0546-patterns/patterns-feature-specification.md [hunk 4]
```diff
@@ -1940,6 +1968,15 @@ To match a pattern `p` against a value `v`:
         method of the same name on the matched value, with `c` as the argument
         returns true.
 
+*   **Cast**:
+
+    1.  If the runtime type of `v` is not a subtype of the cast type of `p` then
+        throw a runtime exception. *Note that we throw even if this appears in a
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1974]: exception -> error. (It'll be a `TypeError`)

File: working/0546-patterns/patterns-feature-specification.md [hunk 5]
```diff
@@ -1949,31 +1986,27 @@ To match a pattern `p` against a value `v`:
 *   **Null-assert**:
 
     1.  If `v` is null then throw a runtime exception. *Note that we throw even
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1988]: exceptionn -> error

File: working/0546-patterns/patterns-feature-specification.md [hunk 6]
```diff
@@ -1949,31 +1986,27 @@ To match a pattern `p` against a value `v`:
...
 
-*   **Cast**:
+*   **Variable**:
 
-    1.  If the runtime type of `v` is not a subtype of the static type of `p`
-        then throw a runtime exception. *Note that we throw even if this appears
-        in a refutable context. The intent of this pattern is to assert that a
-        value *must* have some type.*
+    1.  If the runtime type of `v` is not a subtype of the static type of the
+        variable that `p` refers to then the match fails.
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:2007]: This is an assignment, it can only happen in an irrefutable context.
So, instead of the match failing, we should throw a runtime error (and it will always be a failed downcast from `dynamic`).

**Review by lrhn — APPROVED** [review]

File: working/0546-patterns/patterns-feature-specification.md [hunk 7]
```diff
@@ -1133,35 +1101,153 @@ if (var [int x, int y] = json) {
...
+    const c = 1;
+    switch (2) {
+      case c: print('match $c');
+      default: print('no match');
+    }
+    ```
+
+    *This program prints "no match" and not "match 2".*
+
+*   A simple identifier in a matching context named `_` is treated as a wildcard
```
Comment (stereotype441) [inline working/0546-patterns/patterns-feature-specification.md:1231]: Technically this is a breaking change, since in Dart today we can do:

```dart
main() {
  const _ = 1;
  switch (2) {
    case _:
      print('Matched _');
      break;
    default:
      print('Did not match _');
      break;
  }
}
```

(and this program prints `Did not match _`).

I think this is fine, because the chances of the breakage affecting a real user are so low, but we should probably mention in this doc that it's a breakage just for completeness.

File: working/0546-patterns/patterns-feature-specification.md [hunk 8]
```diff
@@ -1314,12 +1402,15 @@ The context type schema for a pattern `p` is:
...
 
-*   **Literal** or **constant**: The context type schema is the static type of
-    the pattern's constant value expression.
+*   **Constant**: The context type schema is the static type of the pattern's
+    value.
 
 *   **Variable**:
 
-    1.  If `p` has no type annotation, the context type schema is `?`.
+    1.  In an assignment context, the context type schema is the static type of
```
Comment (stereotype441) [inline working/0546-patterns/patterns-feature-specification.md:1410]: What if the variable is promoted?  Does `static type` mean the promoted type or the unpromoted type of the variable?

I think it should mean the unpromoted type, so that demoting assignments are possible (as they are for ordinary non-pattern assignments).

**Review by stereotype441 — APPROVED** [review]
lgtm, though my review is nowhere near as thorough as @lrhn's, so if he contradicts me I defer to him :)

File: working/0546-patterns/patterns-feature-specification.md [hunk 9]
```diff
@@ -144,8 +144,8 @@ matching constructs to make working with that style enjoyable:
 ```dart
 double calculateArea(Shape shape) =>
   switch (shape) {
-    case Square(length: l) => l * l;
-    case Circle(radius: r) => math.pi * r * r;
+    case Square(length: var l) => l * l;
```
Comment (jakemac53) [inline working/0546-patterns/patterns-feature-specification.md:147]: 🎉 this makes things way more clear imo

File: working/0546-patterns/patterns-feature-specification.md [hunk 10]
```diff
@@ -357,10 +343,38 @@ String asciiCharType(int char) {
...
+A cast pattern is similar to an extractor pattern in that it checks the matched
+value against a given type. But where an extractor pattern is *refuted* if the
+value doesn't have that type, a cast pattern *throws*. Like the null-assert
+pattern, this lets you forcibly assert the expected type of some destructured
+value. This isn't useful as the outermost pattern in a declaration since you can
+always move the `as` to the initializer expression:
+
+```dart
+num n = 1;
+var i as int = n; // Instead of this...
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:361]: Just removed this example.

File: working/0546-patterns/patterns-feature-specification.md [hunk 11]
```diff
@@ -671,31 +634,37 @@ Field subpatterns can be in one of three forms:
...
-    variable with the same name. As a convenience, the identifier can be omitted
-    on a named field. In that case, the name is inferred from `pattern`. The
-    subpattern must be a variable pattern or cast pattern, which may be wrapped
-    in any number of null-check or null-assert patterns.
+    variable with the same name.
 
-    The field name is then inferred from the name in the variable or cast
+    As a convenience, the identifier can be omitted and inferred from `pattern`.
+    The subpattern must be a variable pattern which may be wrapped in a unary
+    pattern. The field name is then inferred from the name in the variable
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:646]: More variants might be useful, but I think they quickly get unreadable. I prefer to not support:

```dart
Foo(:var point & Point(:var x,:var y))
```

At some point, it's better to just be clear and write the field name, and I think using `&` and `|` patterns has crossed that threshold.

File: working/0546-patterns/patterns-feature-specification.md [hunk 12]
```diff
@@ -357,10 +343,38 @@ String asciiCharType(int char) {
...
+value. This isn't useful as the outermost pattern in a declaration since you can
+always move the `as` to the initializer expression:
+
+```dart
+num n = 1;
+var i as int = n; // Instead of this...
+var i = n as int; // ...do this.
+```
+
+But when destructuring, there is no place in the initializer to insert the cast.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:365]: Removed the example so I think it flows better with "but" now. :)

File: working/0546-patterns/patterns-feature-specification.md [hunk 13]
```diff
@@ -799,8 +767,7 @@ Add this new rule:
 ```
 patternDeclaration  ::= ( 'final' | 'var' ) outerPattern '=' expression
 
-outerPattern        ::= nullCheckPattern
-                      | parenthesizedPattern
+outerPattern        ::= parenthesizedPattern
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:770]: I don't know if I had a super concrete use case in mind, but yes, that's a valid one. Back when `outerPattern` was also used by `if (var <pattern> = <expr>)`, it was also an escape hatch to let you access refutable patterns at the outermost level.

Your example makes me wonder if we should just allow an unparenthesized `&` in `outerPattern` too. Filed: https://github.com/dart-lang/language/issues/2481

File: working/0546-patterns/patterns-feature-specification.md [hunk 14]
```diff
@@ -811,11 +778,10 @@ The `outerPattern` rule defines a subset of the patterns that are allowed as the
...
 ```dart
-var [a, b] = [1, 2];                  // List.
-var {1: a} = {1: 2};                  // Map.
-var (a, b, x: x) = (1, 2, x: 3);      // Record.
-var Point(x: x, y: y) = Point(1, 2);  // Extractor.
-if (var string? = nullableString) ... // Null-check.
+var [a, b] = [1, 2];                      // List.
+var {1: a} = {1: 2};                      // Map.
+var (a, b, x: x) = (1, 2, x: 3);          // Record.
+var Point(x: x, y: y) = Point(1, 2);      // Extractor.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:784]: Done.

File: working/0546-patterns/patterns-feature-specification.md [hunk 15]
```diff
@@ -848,9 +814,9 @@ forLoopParts ::=
 As with regular for-in loops, it is a compile-time error if the type of
 `expression` in a pattern-for-in loop is not assignable to `Iterable<dynamic>`.
 
-*This could potentially be extended to allow patterns in top-level variables and
-static fields but laziness makes that more complex. We could support patterns in
-instance field declarations, but constructor initializer lists make that harder.
+*We could potentially allow patterns in top-level variables and static fields
+but laziness makes that more complex. We could support patterns in instance
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:818]: Done. :)

File: working/0546-patterns/patterns-feature-specification.md [hunk 16]
```diff
@@ -183,12 +183,11 @@ Before introducing each pattern in detail, here is a summary with some examples:
...
 | [Relational][relationalPattern] | `== expression`<br>`< expression` |
+| [Cast][castPattern] | `foo as String` |
 | [Null-check][nullCheckPattern] | `subpattern?` |
 | [Null-assert][nullAssertPattern] | `subpattern!` |
-| [Literal][literalPattern] | `123`, `null`, `'string'` |
-| [Constant][constantPattern] | `math.pi`, `SomeClass.constant` |
-| [Variable][variablePattern] | `foo`, `String str`, `_`, `int _` |
-| [Cast][castPattern] | `foo as String` |
+| [Constant][constantPattern] | `123`, `null`, `'string'`<br>`math.pi`, `SomeClass.constant`<br>`const Thing(1, 2)`, `const (1 + 2)` |
+| [Variable][variablePattern] | `foo`, `var bar`, `String str`, `_`, `int _` |
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:190]: I went back and forth on this several times, but ultimately I think having a single pattern that covers all of these was the simplest answer.

I could split out wildcards, but I still think it's important to support typed wildcards like `int _`. Those should behave like variable patterns in that they are only allowed in irrefutable contexts if the variable isn't a downcast (which makes them useless, but still), and do a type test when used in a refutable context. The only difference between them and variable patterns is that they don't bind. If I treated `type? '_'` as a separate pattern, I'd have to specify that ambiguity between those and typed variables is resolved to a wildcard (since `_` is a valid `identifier`). Also, `var _` should be non-binding to be consistent with *non*-pattern variable declarations, assuming we can pull off that change too. It just ended up being simpler to treat them as variable patterns.

Likewise, separating out identifiers into their own patterns doesn't buy much. If we use those to refer to named constants in refutable contexts, then the specification of them has to basically copy/paste the specification for constants but only when in a refutable context.

File: working/0546-patterns/patterns-feature-specification.md [hunk 17]
```diff
@@ -872,26 +838,28 @@ patternAssignment ::= outerPattern '=' expression
 assignments, but does not allow patterns to the left of a compound assignment
 operator.*
 
-In a pattern assignment, all variable and cast patterns are interpreted as
-referring to existing variables or setters. You can't declare any new variables.
-*Disallowing new variables allows pattern assignment expressions to appear
-anywhere expressions are allowed while avoiding confusion about the scope of new
-variables.*
+In a pattern assignment, all variable patterns are interpreted as referring to
+existing variables. You can't declare any new variables. *Disallowing new
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:842]: > That's a confusing phrasing when "variable pattern" includes `var x`.

It does include that, but that specific gets ruled out later when talking about context.

File: working/0546-patterns/patterns-feature-specification.md [hunk 18]
```diff
@@ -872,26 +838,28 @@ patternAssignment ::= outerPattern '=' expression
...
+variables allows pattern assignment expressions to appear anywhere expressions
+are allowed while avoiding confusion about the scope of new variables.*
 
 It is a compile-time error if:
 
-*   A variable pattern has a type annotation. *Only simple identifiers are
-    allowed since you aren't declaring a new variable.*
-
 *   An identifier in a variable or cast pattern does not resolve to a non-final
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:848]: I think the way it's specified makes sense here given the other spec choices where variable pattern also includes simple identifiers.

File: working/0546-patterns/patterns-feature-specification.md [hunk 19]
```diff
@@ -1081,11 +1049,11 @@ expression.
 ```dart
 main() {
   switch (some(extremely, long, expression, here)) {
-    case Some(Quite(var long, var pattern)) => expression();
+    case Some(Quite(:var long, :var pattern)) => expression();
   };
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1053]: That's an interesting point. Lambdas already do too (modulo `await`/`yield`/`break`/`continue`). But, yes, switch expressions give you full expression locals even inside loops, async functions, and generators.

Whether that suggests we should do `let` expressions or some other syntax is a separate discussion. :)

File: working/0546-patterns/patterns-feature-specification.md [hunk 20]
```diff
@@ -994,6 +959,9 @@ switch (1) {
 }
 ```
 
+To have an empty case that does *not* fallthrough, use `break;` for its body as
+you would today.
+
 ### Switch expression
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:965]: Probably. Filed: https://github.com/dart-lang/language/issues/2482

File: working/0546-patterns/patterns-feature-specification.md [hunk 21]
```diff
@@ -1133,35 +1101,153 @@ if (var [int x, int y] = json) {
...
+
+    *Logical-or patterns are refutable because there is no point in using one
+    with an irrefutable left operand. We could make null-check patterns
+    irrefutable if `V` is assignable to its static type, but whenever that is
+    true the pattern does nothing useful since its only behavior is a type
+    test.*
+
+    *The remaining patterns are allowed syntactically to appear in a refutable
+    context. Patterns that do type tests like variables and lists produce a
+    compile-time error when used in an irrefutable context if the matched value
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1172]: Good catch. :)

File: working/0546-patterns/patterns-feature-specification.md [hunk 22]
```diff
@@ -1133,35 +1101,153 @@ if (var [int x, int y] = json) {
...
+    *This program prints "no match" and not "match 2".*
+
+*   A simple identifier in a matching context named `_` is treated as a wildcard
+    variable pattern. *A bare `_` is always treated as a wildcard regardless of
+    context, even though other variables in matching contexts require a marker.*
+
+    ```dart
+    // OK:
+    switch (triple) {
+      case [_, y, _]: print('The middle element is $y');
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1238]: Oops, forgot to update this to the new syntax. :) Should be `var y`. Fixed.

File: working/0546-patterns/patterns-feature-specification.md [hunk 23]
```diff
@@ -1133,35 +1101,153 @@ if (var [int x, int y] = json) {
...
+    }
+    ```
+
+    *You can also use `var _` or `final _` to write a wildcard in a matching
+    context because it would require additional specification to explicitly
+    forbid it, but doing so is discouraged.*
+
+*In short, don't use refutable patterns in places that don't do control flow.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1246]: Done.

File: working/0546-patterns/patterns-feature-specification.md [hunk 24]
```diff
@@ -1133,35 +1101,153 @@ if (var [int x, int y] = json) {
...
+Use simple identifiers (optionally with type annotations) to declare variables
+in pattern declarations. Use simple identifiers to assign to variables in
+pattern assignments. Use explicitly marked identifiers to declare variables in
+cases. Use `_` anywhere for a wildcard.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1250]: Done (well, "`case` patterns"`)

File: working/0546-patterns/patterns-feature-specification.md [hunk 25]
```diff
@@ -1314,12 +1402,15 @@ The context type schema for a pattern `p` is:
...
-    the pattern's constant value expression.
+*   **Constant**: The context type schema is the static type of the pattern's
+    value.
 
 *   **Variable**:
 
-    1.  If `p` has no type annotation, the context type schema is `?`.
+    1.  In an assignment context, the context type schema is the static type of
+        the variable that `p` resolves to.
+
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1412]: Yes. A `variablePattern` always refers to a variable and never a constant. A `constantPattern` always refers to a constant. Those two kinds of patterns are syntactically ambiguous since a short identifier is valid for both, but we resolve that syntactically using the surrounding pattern context which is known at parse time (which is why I moved the "pattern context" section up before the static semantics).

File: working/0546-patterns/patterns-feature-specification.md [hunk 26]
```diff
@@ -1404,22 +1496,30 @@ To type check a pattern `p` being matched against a value of type `M`:
 *   **Relational**: If the operator is a comparison (`<`, `<=`, `>`, or `>=`),
     then it is a compile-time error if `M` does not define that operator, if the
     type of the constant in the relational pattern is not assignable to the
-    operator's parameter type, or if the operator's return type is not `bool` or
-    `dynamic`. *The `==` and `!=` operators are valid for all pairs of types.*
+    operator's parameter type, or if the operator's return type is not
+    assignable to `bool`. *The `==` and `!=` operators are valid for all pairs
+    of types.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1501]: Nice catch. Done.

File: working/0546-patterns/patterns-feature-specification.md [hunk 27]
```diff
@@ -1404,22 +1496,30 @@ To type check a pattern `p` being matched against a value of type `M`:
...
+
+*   **Cast**:
+
+    1.  Resolve the type name to a type `X`. It is a compile-time error if
+        the name does not refer to a type.
+
+        **TODO: Should we infer type arguments on `X`?**
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1508]: SGTM.

File: working/0546-patterns/patterns-feature-specification.md [hunk 28]
```diff
@@ -1264,14 +1350,16 @@ To orchestrate this, type inference on patterns proceeds in three phases:
     holes in the type schema. When that completes, we now have a full static
     type for the pattern and all of its subpatterns.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1351]: Exactly right, yes.

File: working/0546-patterns/patterns-feature-specification.md [hunk 29]
```diff
@@ -1537,130 +1639,46 @@ To type check a pattern `p` being matched against a value of type `M`:
...
-
-In the first function, the `(a, b)` pattern will always successfully destructure
-the record because `obj` is known to be a record type of the right shape. But in
-the second function, `obj` may fail to match because the value may not be a
-record. *This implies that we can't determine whether a pattern in a variable
-declaration is incorrectly refutable until after type checking.*
+It is a compile-time error if:
 
-Refutability of a pattern `p` matching a value of type `V` is:
+*   The type of an expression in a guard clause is not `bool` or `dynamic`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1646]: Oops, yes. Good catch. :)

File: working/0546-patterns/patterns-feature-specification.md [hunk 30]
```diff
@@ -1537,130 +1639,46 @@ To type check a pattern `p` being matched against a value of type `M`:
...
 
-*   **Logical-and**, **parenthesized**, **null-assert**, or **cast**:
-    Irrefutable if and only if all subpatterns are irrefutable.
+*   `p` is in an irrefutable context, it has a required type `T`, and `M` is not
+    assignable to `T`. *Destructuring and variable patterns can only be used in
+    declarations and assignments if we can statically tell that the
+    destructuring and variable binding won't fail.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1651]: "fail" here means "fail to match" and not "complete successfully". Reworded and added a parenthetical to clarify.

File: working/0546-patterns/patterns-feature-specification.md [hunk 31]
```diff
@@ -1537,130 +1639,46 @@ To type check a pattern `p` being matched against a value of type `M`:
...
 Patterns often exist to introduce new variable bindings. A "wildcard" identifier
-named `_` in a variable or cast pattern never introduces a binding.
-
+named `_` in a variable pattern never introduces a binding.
 The variables a patterns binds depend on what kind of pattern it is:
 
 *   **Logical-or**: Does not introduce variables but may contain subpatterns
-    that do. If it a compile-time error if the two subpatterns do not introduce
+    that do. It is a compile-time error if the two subpatterns do not introduce
     the same variables with the same names and types.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1669]: Yeah, I agree this is definitely too vague. For now, I'm going to just put a "TODO" here, because I don't want to bother specifying something in detail that will then likely get replaced when I address https://github.com/dart-lang/language/issues/2473.

File: working/0546-patterns/patterns-feature-specification.md (hunk 1 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1909]: I think the "Variables and scope" section covers this pretty well, but if you think it's better to explain it here I could do that.

File: working/0546-patterns/patterns-feature-specification.md (hunk 2 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1914]: Done.

File: working/0546-patterns/patterns-feature-specification.md (hunk 3 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1917]: I like that. Done.

File: working/0546-patterns/patterns-feature-specification.md (hunk 4 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1974]: Done.

File: working/0546-patterns/patterns-feature-specification.md (hunk 5 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1988]: Done (here and elsewhere).

File: working/0546-patterns/patterns-feature-specification.md (hunk 6 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:2007]: The sentence isn't clear, but the intent is that "the variable that `p` refers to" can also mean "the variable that `p` itself is declaring". So it can occur in a refutable context in which case it is a type test:

```dart
if ((123 as Object) case String s) { ... }
```

I'll reword this to make it clearer.

File: working/0546-patterns/patterns-feature-specification.md (hunk 7 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1231]: Good point. I went ahead and consolidated all of the breaking stuff into a new section.

File: working/0546-patterns/patterns-feature-specification.md (hunk 9 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:147]: I think so too, but I really appreciate the feedback. We went back and forth on this *so* much. :)

File: working/0546-patterns/patterns-feature-specification.md (hunk 8 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1410]: Haha, good question! I'll defer to your expertise. Done. :)

File: working/0546-patterns/patterns-feature-specification.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply working/0546-patterns/patterns-feature-specification.md:190]: If people can read it and understand which identifier means what where, then it works. Ship it.

It'll be Erik's problem to do the final formal specification anyway :smiling_imp:

File: working/0546-patterns/patterns-feature-specification.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply working/0546-patterns/patterns-feature-specification.md:613]: ACK.

File: working/0546-patterns/patterns-feature-specification.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply working/0546-patterns/patterns-feature-specification.md:1118]: We need to be very clear on the scoping of such variables. Each evaluation must introduce a new binding since the variable can be final. That's *probably* ensured by putting the variable in the nearest enclosing scope, because there is no way to evaluate the same expression twice (aka. loop) without introducing a new scope per iteration. We just need to make sure we haven't forgotten (or introduced) any operation where that's not true.

For switch cases, we need a scope per `casehead`, because each can introduce independent final variables that can be read in the `when` clause, whether the case is then accepted or not.

Not a problem, we just need to be careful.

File: working/0546-patterns/patterns-feature-specification.md (hunk 8 above)
Comment (lrhn) [reply working/0546-patterns/patterns-feature-specification.md:1475]: Do what we'd do on a plain assignment, `x = e`, for the same variable.

File: working/0546-patterns/patterns-feature-specification.md [hunk 32]
```diff
@@ -1374,12 +1467,15 @@ The context type schema for a pattern `p` is:
...
-*   **Literal** or **constant**: The context type schema is the static type of
-    the pattern's constant value expression.
+*   **Constant**: The context type schema is the static type of the pattern's
+    value.
 
 *   **Variable**:
 
-    1.  If `p` has no type annotation, the context type schema is `?`.
+    1.  In an assignment context, the context type schema is the static type of
+        the variable that `p` resolves to.
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1476]: (Should we disallow assigning the same variable more than once in the same assignment pattern?)

File: working/0546-patterns/patterns-feature-specification.md (hunk 1 above)
Comment (lrhn) [reply working/0546-patterns/patterns-feature-specification.md:1990]: \*\* goes to read "Variables and scope"\*\*

File: working/0546-patterns/patterns-feature-specification.md [hunk 33]
```diff
@@ -1600,44 +1718,51 @@ To type check a pattern `p` being matched against a value of type `M`:
...
-    not final.*
+[2473]: https://github.com/dart-lang/language/issues/2473
 
-The scope where a pattern's variables are declared depends on the construct
-that contains the pattern:
+The scope where a pattern's variables are declared and available for use depends
+on the construct that contains the pattern:
 
 *   **Local pattern variable declaration**: The rest of the block following
     the declaration.
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1768]: This is new. The current behavior is that the scope is the entire block, and access only allowed in the rest of the block.
It matters for 
```dart
var x = 42;
{
  var y = x;  // <- invalid access to inner x
  var x = y;
}
```
We should probably do the same thing for pattern and non-pattern declarations.

File: working/0546-patterns/patterns-feature-specification.md [hunk 34]
```diff
@@ -1600,44 +1718,51 @@ To type check a pattern `p` being matched against a value of type `M`:
...
 
-*   **Variable** or **cast**: May contain type argument patterns. Introduces a
-    variable whose name is the pattern's identifier. The variable is final if
-    the surrounding `patternDeclaration` has a `final` modifier. *Variables in
-    switch cases don't occur inside pattern variable declarations and thus are
-    not final.*
+[2473]: https://github.com/dart-lang/language/issues/2473
 
-The scope where a pattern's variables are declared depends on the construct
-that contains the pattern:
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1640]: The scope where a local variable is declared, and where it's available for use, are not the same thing in the current specification. We need to define both.

File: working/0546-patterns/patterns-feature-specification.md [hunk 35]
```diff
@@ -1646,7 +1771,7 @@ that contains the pattern:
 *   **Switch statement case**: The guard clause and the statements of the
     subsequent non-empty case body.
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1772]: If multiple `casehead`s can reach the same body, then we need some way to merge scopes.

Take:
```dart
 case Point(:int x, :int y):
 case (:int x, :int y):
    x += 1;
    y += 1;
    print(x + y);
```
Both branches introduce the same names with the same types. One of them reaches the body, and defines the variables that are modified. But *both* sets of variables can be in play:

```dart
late int Function() next;
...
 case Point(:int x, :int y) when (next = () => x++) != 0 && x < y:
 case Point(:int x, :int y): 
    print([x++, y++, next());
```
If given `Point(x:10, y: 5)` the first case will be rejected in the `when` clause, but still capture the `x` variable of that.
Then the next case succeeds, binding a new `x` and `y`.
It should print `[10, 5, 10]`.


Is this just an argument against allowing multiple `casehead`s for the same body? I think it is!

You can still capture variables from *other* cases, which were rejected in the `when` clauses, but those will definitely be *different* variables, but you only get one pattern scope and one `when` clause per body.

```dart
late int Function() next;
...
 case Point(:int x, :int y) when (next = () => x++) != 0 && false:
    print("not today");
 case Point(:int x, :int y): 
    print([x++, y++, next());
```
This still works and binds the variable of the first (matching!) pattern, but there is no way for those variables to ever flow into the same body as the second `case`.

We can allow `case c1: case c2: case c3: body` as a backwards compatible way to write `case c1 | c2 | c3: body`, but you can't add bindings or `when` clauses to any of those cases. To do that, you have to combine them to one pattern using `|`.

On the other hand, that would prohibit:
```dart
case Value(:var x) when x >= 0:
case NegatedValue(:var x) when x < 0:
  same body;
```
if that's something you want to do (multiple independent `when`s leading to the same body).

File: working/0546-patterns/patterns-feature-specification.md [hunk 36]
```diff
@@ -1646,7 +1771,7 @@ that contains the pattern:
...
 *   **Switch expression case**: The guard clause and the case expression.
-*   **Pattern-if statement**: The then statement.
+*   **If-case statement**: The guard clause and then statement.
 
 Multiple switch case patterns may share the same variable scope if their case
 bodies are empty:
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1777]: Nope. Just thought about this, and it *cannot* work.

```dart
late int Function() otther;
...
switch([1, 2]) {
  case [final int a, final int b] when (other = ()=>a) == 0:
  case [final int b, final int a]:
     print(a + b + other(); // 2 + 1 + 1
```

The scope of each `casehead` must be independent, because it can bind *final* variables to values even if then refuted by a `when` clause. The same final variable cannot be initialized more than once. 
And since a `when` clause is an expression, it can capture the final variable, and must see the correct value.

So each `case` pattern introduces its own bindings.

We then have to make *one* of those sets of bindings the current binding for the body, when we execute/evaluate that body. That's a little weird, but not much weirder than what we do for `for(var x = 1;x < 10;x++)` iterations. (Which isn't really saying much. That *is* weird.)

See more above.

File: working/0546-patterns/patterns-feature-specification.md [hunk 37]
```diff
@@ -1668,7 +1793,7 @@ However, it is a compile-time error if all switch case patterns that share a
 body do not all define the exact same variables with the exact same types.
 
 *Aside from this special case, note that since all variables declared by a
-pattern and its subpattern go into the same scope, it is an error if two
+pattern and its subpatterns go into the same scope, it is an error if two
 subpatterns declare a variable with the same name, unless the name is `_`.*
```
Comment (lrhn) [inline working/0546-patterns/patterns-feature-specification.md:1797]: Even if those are sub-patterns of a `|` pattern?

File: working/0546-patterns/patterns-feature-specification.md (hunk 32 above)
Comment (munificent) [reply working/0546-patterns/patterns-feature-specification.md:1476]: Yes, that's covered in the initial syntax section for pattern assignments.
